crates/web-sys-async-io/src/reader.rs

```diff
@@ -68,17 +57,20 @@ impl tokio::io::AsyncRead for Reader {
                 Poll::Ready(Ok(()))
             }
             Op::Idle => {
-                let work_buf = match &mut self.internal_buf {
-                    Some(val) => val.buffer(),
-                    None => {
-                        let internal_buf =
-                            js_sys::Uint8Array::new_with_length(buf.capacity().try_into().unwrap());
-                        let view = internal_buf.buffer();
-                        self.internal_buf = Some(internal_buf);
-                        view
-                    }
-                };
-                let fut = JsFuture::from(self.inner.read_with_array_buffer_view(&work_buf));
+                let desired = buf.remaining() as u32;
+
+                let mut internal_buf = self
+                    .internal_buf
+                    .get_or_insert_with(|| js_sys::ArrayBuffer::new(desired));
+
+                if internal_buf.byte_length() < desired {
+                    internal_buf = self.internal_buf.insert(js_sys::ArrayBuffer::new(desired));
+                }
+
+                let view =
+                    js_sys::Uint8Array::new_with_byte_offset_and_length(&internal_buf, 0, desired);
```

Comment:
This kind of defeats the purpose of having an internal buffer - which is to avoid unneeded allocations

Reply:
The actual storage would be in the ArrayBuffer, and we need a separate Uint8Array to be able to set the size